File: src/main/java/catmoe/fallencrystal/limborule/command/commands/SendLimbo.java
package catmoe.fallencrystal.limborule.command.commands;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import catmoe.fallencrystal.limborule.command.SubCommand;
import catmoe.fallencrystal.limborule.util.LimboCreater;
import catmoe.fallencrystal.limborule.util.MessageUtil;
import net.md_5.bungee.api.CommandSender;
import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.connection.ProxiedPlayer;

public class SendLimbo implements SubCommand {

    @Override
    public void execute(CommandSender sender, String[] args) {
        // 创建一个LimboCreater
        LimboCreater limbo = new LimboCreater();
        // 将自己发送到Limbo
        if (args[1].equalsIgnoreCase("me")) {
            if (!(sender instanceof ProxiedPlayer)) {
                MessageUtil.prefixsender(sender, "&cConsole is a invalid target.");
                return;
            }
            limbo.CreateServer((ProxiedPlayer) sender);
            limbo.Connect((ProxiedPlayer) sender);
        }
        if (args[1].equalsIgnoreCase("all")) {
            for (ProxiedPlayer p : ProxyServer.getInstance().getPlayers()) {
                limbo.CreateServer(p);
                limbo.Connect(p);
            }
        } else {
            ProxiedPlayer p = null;
            // 判定是否有这个玩家 如果非玩家通常会抛出错误
            try {
                p = ProxyServer.getInstance().getPlayer(args[1]);
            } catch (Exception e) {
                return; // 非玩家
            }
            limbo.CreateServer(p);
            limbo.Connect(p);
        }
    }

    @Override
    public String getSubCommandId() {
        return "sendlimbo";
    }

    @Override
    public String getPermission() {
        return "akanelimbo.sendlimbo";
    }

    @Override
    public Map<Integer, List<String>> getTabCompleter() {
        Map<Integer, List<String>> map = new HashMap<>();
        ArrayList<String> Tip1 = new ArrayList<>();
        for (ProxiedPlayer p : ProxyServer.getInstance().getPlayers()) {
            Tip1.add(p.toString());
        }
        Tip1.add("me");
        Tip1.add("all");
        map.put(1, Tip1);
        return map;
    }

    @Override
    public boolean allowedConsole() {
        return true;
    }

    @Override
    public boolean StrictSizeLimit() {
        return true;
    }

    @Override
    public int StrictSize() {
        return 3;
    }

}

File: src/main/java/catmoe/fallencrystal/akanelimbo/command/commands/SendLimbo.kt
package catmoe.fallencrystal.akanelimbo.command.commands

import catmoe.fallencrystal.akanelimbo.StringManager
import catmoe.fallencrystal.akanelimbo.command.SubCommand
import catmoe.fallencrystal.akanelimbo.util.LimboCreater
import catmoe.fallencrystal.akanelimbo.util.MessageUtil.prefixsender
import net.md_5.bungee.api.CommandSender
import net.md_5.bungee.api.ProxyServer
import net.md_5.bungee.api.connection.ProxiedPlayer

class SendLimbo : SubCommand {
    override val subCommandId: String
        get() = "sendlimbo"

    override fun execute(sender: CommandSender?, args: Array<String>) {
        // 创建一个LimboCreater
        val limbo = LimboCreater()
        // 将自己发送到Limbo
        if (args[1].equals("me", ignoreCase = true)) {
            if (sender !is ProxiedPlayer) {
                prefixsender(sender!!, "&cConsole is a invalid target.")
                return
            }
            limbo.createServer(sender, StringManager.getCommandLimbo())
            limbo.connect(sender)
        }
        if (args[1].equals("all", ignoreCase = true)) {
            for (p in ProxyServer.getInstance().players) {
                limbo.createServer(p!!, StringManager.getCommandLimbo())
                limbo.connect(p)
            }
        } else {
            try {
                val p = ProxyServer.getInstance().getPlayer(args[1])
                if (p != null) {
                    limbo.createServer(p, StringManager.getCommandLimbo())
                    limbo.connect(p)
                }
            } catch (ignore: Exception) {
            }
        }
    }

    override val permission: String
        get() = StringManager.getSendLimboPermission()

    override val tabCompleter: MutableMap<Int, List<String>>

        get() = getTabCompleterMap()

    private fun getTabCompleterMap(): MutableMap<Int, List<String>> {
        val map: MutableMap<Int, List<String>> = HashMap()
        val tip1 = ArrayList<String>()
        for (p in ProxyServer.getInstance().players) {
            tip1.add(p.toString())
        }
        tip1.add("me")
        tip1.add("all")
        map[1] = tip1
        return map
    }

    override fun allowedConsole(): Boolean {
        return true
    }

    override fun strictSizeLimit(): Boolean {
        return true
    }

    override fun strictSize(): Int {
        return 2
    }
}
